src/OrchardCore.Modules/OrchardCore.Tenants/Services/SaasTokenManager.cs

```diff
@@ -0,0 +1,15 @@
+using System;
+using System.Collections.Generic;
+using System.Text;
+using OrchardCore.Tenants.Abstractions;
+
+namespace OrchardCore.Tenants.Services
+{
+    class SaasTokenManager : ISaasTokenManager
```

Comment:
It would be nice improvement, but In this case, we have to inject ISaasTokenManager to the SetupController, which currently is not possible, while the SaasTokenManager is located in Tenants module and we don't have access to that in Setup module

To me, one solution is refactoring the ISaasTokenManager and move it to a new Module which can be used by both Tenants and Setup modules. Do you have a better suggestion?

Reply:
Try to resolve the token manager from the Default host container.

Reply:
- In the site settings, you store a random string (the secret cryptographically random) in plain text
- The url is generated by encrypting the secret with a time limit (2 days). -> byte[] -> Base64
- From the setup controller you read the encrypted secret, decrypt, check the decrypted value is the same as the one in the site settings, and that the time limit is still fine, or return 404.

Do the check in both the GET and the POST on the setup controller.
The IDataProtector needs to be resolved from the IShellHost.